File: backend/app/models/deal.py
from datetime import datetime
from typing import Literal
from pydantic import Field, field_validator

from app.models.common import ApiModel
from app.models.feed_provider import FeedProviderAdapter


DealSort = Literal[
    "score_desc",
    "discount_desc",
    "price_asc",
    "price_desc",
    "rating_desc",
    "newest",
]

DealMonetizationMode = Literal[
    "affiliate",
    "direct",
    "pending_affiliate",
]

DealDeliveryRegion = Literal[
    "global",
    "cis",
    "europe",
    "usa",
    "latam",
]


class Deal(ApiModel):
    id: str
    title: str
    description: str
    image_url: str
    platform: str
    category: str
    old_price: float
    current_price: float
    currency: str = "USD"
    product_url: str
    affiliate_url: str | None = None
    provider_id: str | None = None
    monetization_mode: DealMonetizationMode = "direct"
    rating: float = Field(ge=0, le=5)
    review_count: int = 0
    free_shipping: bool = False
    verified: bool = False
    ships_to: list[str] = Field(default_factory=list)
    delivery_regions: list[DealDeliveryRegion] = Field(default_factory=list)
    hot_deal: bool = False
    lowest_price: bool = False
    deal_score: int = Field(default=0, ge=0, le=100)
    updated_at: datetime
    expires_at: datetime | None = None

    @property
    def discount_percent(self) -> int:
        if self.old_price <= 0 or self.current_price <= 0 or self.old_price <= self.current_price:
            return 0
        percent = round(((self.old_price - self.current_price) / self.old_price) * 100)
        if percent < 1:
            return 0
        return min(percent, 100)


class DealResponse(ApiModel):
    id: str
    title: str
    description: str
    image_url: str
    platform: str
    category: str
    old_price: float
    current_price: float
    currency: str
    product_url: str
    affiliate_url: str | None = None
    provider_id: str | None = None
    monetization_mode: DealMonetizationMode = "direct"
    rating: float
    review_count: int
    free_shipping: bool
    verified: bool
    ships_to: list[str]
    delivery_regions: list[DealDeliveryRegion]
    hot_deal: bool
    lowest_price: bool
    deal_score: int
    discount_percent: int
    updated_at: datetime
    expires_at: datetime | None = None


class DealsPage(ApiModel):
    items: list[DealResponse]
    total: int
    page: int
    page_size: int
    has_next_page: bool


class DealFacetItem(ApiModel):
    id: str
    name: str
    count: int


class DealPriceRange(ApiModel):
    min: float | None = None
    max: float | None = None
    currency: str = "USD"


class DealDiscountRange(ApiModel):
    min: int | None = None
    max: int | None = None


class DealsFacetsResponse(ApiModel):
    total: int
    marketplaces: list[DealFacetItem]
    categories: list[DealFacetItem]
    shipping_countries: list[DealFacetItem]
    delivery_regions: list[DealFacetItem]
    currencies: list[DealFacetItem]
    monetization_modes: list[DealFacetItem]
    price_range: DealPriceRange
    discount_range: DealDiscountRange
    generated_at: datetime


class DealUpsertRequest(ApiModel):
    @field_validator("rating", mode="before")
    @classmethod
    def normalize_rating(cls, value: object) -> float:
        """Accept merchant/feed ratings in 0-5, 0-10, or 0-100 formats.

        Some affiliate feeds, especially Awin product feeds, use `rating` as a
        percentage (80, 95, 100) instead of a 0-5 star value. The public API and
        Flutter filters expect 0-5, so normalize before Pydantic range checks.
        """
        if value is None or value == "":
            return 0
        try:
            if isinstance(value, str):
                cleaned = value.strip().replace("%", "").replace(",", ".")
                rating = float(cleaned)
            else:
                rating = float(value)  # type: ignore[arg-type]
        except (TypeError, ValueError):
            return 0

        if rating < 0:
            return 0
        if rating <= 5:
            return round(rating, 2)
        if rating <= 10:
            return round(rating / 2, 2)
        if rating <= 100:
            return round(rating / 20, 2)
        return 5

    id: str
    title: str
    description: str
    image_url: str
    platform: str
    category: str
    old_price: float = Field(gt=0)
    current_price: float = Field(gt=0)
    currency: str = "USD"
    product_url: str
    affiliate_url: str | None = None
    provider_id: str | None = None
    # None means: infer from affiliate_url. Existing feed adapters do not need to change.
    monetization_mode: DealMonetizationMode | None = None
    rating: float = Field(default=0, ge=0, le=5)
    review_count: int = Field(default=0, ge=0)
    free_shipping: bool = False
    verified: bool = False
    ships_to: list[str] = Field(default_factory=list)
    delivery_regions: list[DealDeliveryRegion] = Field(default_factory=list)
    hot_deal: bool = False
    lowest_price: bool = False
    deal_score: int | None = Field(default=None, ge=0, le=100)
    updated_at: datetime | None = None
    expires_at: datetime | None = None


class BulkDealUpsertRequest(ApiModel):
    items: list[DealUpsertRequest]


class DealsImportRequest(ApiModel):
    items: list[DealUpsertRequest]
    replace: bool = False


class DealsImportUrlRequest(ApiModel):
    url: str
    adapter: FeedProviderAdapter = "auto"
    replace: bool = False
    timeout_seconds: int = Field(default=20, ge=3, le=300)


class DealsExportResponse(ApiModel):
    status: str
    exported_at: datetime
    total: int
    items: list[DealUpsertRequest]


class AdminActionResponse(ApiModel):
    status: str
    message: str
    deal_count: int | None = None


class RatesResponse(ApiModel):
    base_currency: str
    rates: dict[str, float]
    note: str

File: lib/features/deals/models/deal.dart
class Deal {
  const Deal({
    required this.id,
    required this.title,
    required this.description,
    required this.imageUrl,
    required this.platform,
    required this.category,
    required this.oldPrice,
    required this.currentPrice,
    required this.currency,
    required this.productUrl,
    required this.rating,
    required this.reviewCount,
    required this.freeShipping,
    required this.verified,
    required this.shipsTo,
    this.availabilityCountries = const <String>[],
    this.isGlobal = false,
    this.deliveryRegions = const <String>[],
    this.providerId,
    this.monetizationMode = 'direct',
    this.hotDeal = false,
    this.lowestPrice = false,
    this.dealScore = 0,
    this.updatedAt,
  });

  final String id;
  final String title;
  final String description;
  final String imageUrl;
  final String platform;
  final String category;
  final double oldPrice;
  final double currentPrice;
  final String currency;
  final String productUrl;
  final double rating;
  final int reviewCount;
  final bool freeShipping;
  final bool verified;
  final List<String> shipsTo;
  final List<String> availabilityCountries;
  final bool isGlobal;
  final List<String> deliveryRegions;
  final String? providerId;
  final String monetizationMode;
  final bool hotDeal;
  final bool lowestPrice;
  final int dealScore;
  final DateTime? updatedAt;

  factory Deal.fromJson(Map<String, dynamic> json) {
    return Deal(
      id: _string(json['id']),
      title: _string(json['title']),
      description: _string(json['description']),
      imageUrl: _string(json['imageUrl']),
      platform: _string(json['platform']),
      category: _string(json['category']),
      oldPrice: _double(json['oldPrice']),
      currentPrice: _double(json['currentPrice']),
      currency: _string(json['currency'], fallback: 'USD'),
      productUrl: _string(json['productUrl']),
      rating: _double(json['rating']),
      reviewCount: _int(json['reviewCount']),
      freeShipping: _bool(json['freeShipping']),
      verified: _bool(json['verified']),
      shipsTo: _stringList(json['shipsTo']),
      availabilityCountries: _stringList(
        json['availabilityCountries'] ?? json['availability_countries'],
      ),
      isGlobal: _bool(json['isGlobal'] ?? json['is_global']),
      deliveryRegions: _stringList(
        json['deliveryRegions'] ?? json['delivery_regions'],
      ),
      providerId: _nullableString(json['providerId'] ?? json['provider_id']),
      monetizationMode: _string(
        json['monetizationMode'] ?? json['monetization_mode'],
        fallback: 'direct',
      ),
      hotDeal: _bool(json['hotDeal'] ?? json['hot_deal']),
      lowestPrice: _bool(json['lowestPrice'] ?? json['lowest_price']),
      dealScore: _int(json['dealScore'] ?? json['deal_score']),
      updatedAt: _dateTime(json['updatedAt'] ?? json['updated_at']),
    );
  }

  Map<String, dynamic> toJson() {
    return <String, dynamic>{
      'id': id,
      'title': title,
      'description': description,
      'imageUrl': imageUrl,
      'platform': platform,
      'category': category,
      'oldPrice': oldPrice,
      'currentPrice': currentPrice,
      'currency': currency,
      'productUrl': productUrl,
      'rating': rating,
      'reviewCount': reviewCount,
      'freeShipping': freeShipping,
      'verified': verified,
      'shipsTo': shipsTo,
      'availabilityCountries': availabilityCountries,
      'isGlobal': isGlobal,
      'deliveryRegions': deliveryRegions,
      'providerId': providerId,
      'monetizationMode': monetizationMode,
      'hotDeal': hotDeal,
      'lowestPrice': lowestPrice,
      'dealScore': dealScore,
      'updatedAt': updatedAt?.toIso8601String(),
    };
  }

  double get rawDiscountPercent {
    if (oldPrice <= 0 || currentPrice <= 0 || oldPrice <= currentPrice) {
      return 0;
    }
    return ((oldPrice - currentPrice) / oldPrice) * 100;
  }

  bool get hasRealDiscount => rawDiscountPercent >= 1;

  int get discountPercent {
    if (!hasRealDiscount) {
      return 0;
    }
    final percent = rawDiscountPercent.round();
    if (percent < 1) return 0;
    if (percent > 100) return 100;
    return percent;
  }

  double get savedAmount {
    if (!hasRealDiscount) {
      return 0;
    }
    return oldPrice - currentPrice;
  }

  String get formattedCurrentPrice =>
      '$currency ${currentPrice.toStringAsFixed(2)}';

  String get formattedOldPrice => '$currency ${oldPrice.toStringAsFixed(2)}';

  String get formattedSavedAmount =>
      '$currency ${savedAmount.toStringAsFixed(2)}';

  static DateTime? _dateTime(dynamic value) {
    if (value is DateTime) return value;
    if (value is String && value.trim().isNotEmpty) {
      return DateTime.tryParse(value.trim());
    }
    return null;
  }

  static String _string(dynamic value, {String fallback = ''}) {
    if (value is String && value.trim().isNotEmpty) return value.trim();
    if (value is num) return value.toString();
    return fallback;
  }

  static String? _nullableString(dynamic value) {
    if (value is String && value.trim().isNotEmpty) return value.trim();
    if (value is num) return value.toString();
    return null;
  }

  static double _double(dynamic value) {
    if (value is double) return value;
    if (value is int) return value.toDouble();
    if (value is String) return double.tryParse(value) ?? 0;
    return 0;
  }

  static int _int(dynamic value) {
    if (value is int) return value;
    if (value is double) return value.round();
    if (value is String) return int.tryParse(value) ?? 0;
    return 0;
  }

  static bool _bool(dynamic value) {
    if (value is bool) return value;
    if (value is String) return value.toLowerCase() == 'true';
    if (value is num) return value != 0;
    return false;
  }

  static List<String> _stringList(dynamic value) {
    if (value is! List) return const <String>[];

    return value
        .whereType<Object>()
        .map((item) => item.toString().trim())
        .where((item) => item.isNotEmpty)
        .toList(growable: false);
  }
}
